tgs/src/main/cpp/tgs/RStarTree/Box.cpp

```diff
@@ -24,217 +24,224 @@
+  {
+    assert(b.getDimensions() == getDimensions());
+    double result = 1.0;
+    for (int i = 0; i < _dimensions; i++)
+    {
+      double ub = std::min(getUpperBound(i), b.getUpperBound(i));
+      double lb = std::max(getLowerBound(i), b.getLowerBound(i));
+      if (ub - lb <= 0.0)
+      {
+        return 0.0;
+      }
+      result *= ub - lb;
+    }
+    return result;
+  }
+
+  double Box::calculatePerimeter() const
+  {
+    double result = 0.0;
+    for (int i = 0; i < _dimensions; i++)
+    {
+      result += getUpperBound(i) - getLowerBound(i);
+    }
+    return result * 2.0;
+  }
+
+  double Box::calculateVolume() const
+  {
+    double result = 1;
+    for (int i = 0; i < _dimensions; i++)
+    {
+      result *= getUpperBound(i) - getLowerBound(i);
+    }
+    return result;
+  }
+
+  Box& Box::expand(const BoxInternalData& b)
+  {
+    assert(getDimensions() == b.getDimensions());
+    if (isValid() == false)
+    {
+      _valid = true;
+      for (int i = 0; i < getDimensions(); i++)
+      {
+        _lowerBound[i] = b.getLowerBound(i);
+        _upperBound[i] = b.getUpperBound(i);
+      }
+    }
+    else
+    {
+      for (int i = 0; i < getDimensions(); i++)
+      {
+        _lowerBound[i] = std::min(getLowerBound(i), b.getLowerBound(i));
+        _upperBound[i] = std::max(getUpperBound(i), b.getUpperBound(i));
+      }
+    }
+    // for convenience
+    return *this;
+  }
+
+  Box& Box::expand(const Box& b)
+  {
+    assert(getDimensions() == b.getDimensions());
+    if (isValid() == false)
+    {
+      _valid = true;
+      for (int i = 0; i < getDimensions(); i++)
+      {
+        _lowerBound[i] = b.getLowerBound(i);
+        _upperBound[i] = b.getUpperBound(i);
+      }
+    }
+    else
+    {
+      for (int i = 0; i < getDimensions(); i++)
+      {
+        _lowerBound[i] = std::min(getLowerBound(i), b.getLowerBound(i));
+        _upperBound[i] = std::max(getUpperBound(i), b.getUpperBound(i));
+      }
+    }
+    // for convenience
+    return *this;
+  }
+
+  void Box::setBounds(int d, double lower, double upper)
+  {
+    _valid = true;
+    _lowerBound[d] = lower;
+    _upperBound[d] = upper;
+  }
+
+  void Box::setDimensions(int d)
+  {
+    _valid = false;
+    _dimensions = d;
+  }
+
+  double Box::getLowerBound(int d) const
+  {
+    return std::min(_lowerBound[d], _upperBound[d]);
+  }
+
+  double Box::getUpperBound(int d) const
+  {
+    return std::max(_lowerBound[d], _upperBound[d]);
+  }
+
+  double Box::getLowerBoundRaw(int d) const
+  {
+    return _lowerBound[d];
+  }
+
+  double Box::getUpperBoundRaw(int d) const
+  {
+    return _upperBound[d];
+  }
+
+  bool Box::isValid() const
+  {
+    return _valid;
+  }
+
+  bool Box::operator==(const Box& b)
+  {
+    if (_dimensions != b._dimensions)
+    {
+      return false;
+    }
+    for (int i = 0; i < _dimensions; i++)
+    {
+      if (getLowerBound(i) != b.getLowerBound(i) ||
+          getUpperBound(i) != b.getUpperBound(i))
```

Comment:
This used to never return `false` based on the upper bound because it used to compare `b.getUpperBound(i)` with itself.  :logic_bomb: